Refactoring code in `mk_symbol` to remove stock specific code

udesou opened the pull request:
This PR lifts up some code that uses specific information about the stock GC (`GC_OLD_MARKED` in the header bits). MMTk sets the log bits for immortal objects in a side table, but this requires that they are allocated with `jl_gc_permobj` instead of `jl_gc_perm_alloc`, since the latter allocates a memory slot and not necessarily an object.

udesou committed:
Refactoring code in mk_symbol to remove stock specific code

Keno commented on src/gc-interface.h:
```diff
@@ -207,6 +207,13 @@ JL_DLLEXPORT void *jl_gc_perm_alloc(size_t sz, int zero, unsigned align,
+// 2. align for this function is word align, and align for jl_gc_permobj depends on the allocation size.
+// 3. ty for this function is jl_symbol_tag << 4 and is set via jl_set_typetagof, and ty for jl_gc_permobj
+// is a datatype pointer and is set directly from the object's header.
+struct _jl_value_t *jl_gc_permsymbol(size_t sz) JL_NOTSAFEPOINT;
```
I don't really like making this specific to symbol. If the primary difference is in the alignment, why not make a version of permobj that just takes an additional alignment arg?

udesou replied:
I think the alignment and size differences are actually fine, and it would be easy to integrate them with `jl_gc_permobj`. 
The problem is setting the type: because for symbols the type is set `jl_symbol_tag`, the type needs to be set atomically, whereas for objects it can be set directly. I could combine the functions but it would mean that we would have to use `jl_set_typeof` in `jl_gc_permobj`. I think it would look much cleaner but it depends on whether that's a reasonable trade-off.

Keno replied:
> because for symbols the type is set `jl_symbol_tag`, the type needs to be set atomically

This statement doesn't make sense to me.

udesou replied:
Sorry, let me try to clarify. 
The type of the symbol is set [here](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/symbol.c#L42) via a call to [`jl_set_typetagof`](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/julia.h#L137), whereas, for `jl_gc_permobj`, the type is set [here](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/gc-stock.c#L3939) writing it directly to the header. As far as I understood from the code, that difference happens because for symbols we're setting `jl_symbol_tag`, which needs to be set atomically. So if we want to reuse `jl_gc_permobj` to allocate symbols, we need to make sure that it also sets the type atomically.

Keno replied:
I think it's probably just an oversight that the `header` field is not marked atomic.

udesou replied:
I see. I've changed the code to use `jl_gc_permobj` for symbols, and to use `jl_set_typeof` inside it. Thanks for looking into it and for the feedback :)

vchuravy reviewed (commented):
This code still looks a bit confusing to me.

It's overall a bit odd that the smalltag is handled outside of `gc_alloc` e.g. callers have to check `dt->smalltag` and then use `jl_set_typetag_of`.

Maybe `gc_permobj` should take a `jl_value_ty*` and then be passed a `jl_symbol_type`,
so that the caller doesn't have to pre-shift the tag?

udesou commented:
> This code still looks a bit confusing to me.
> 
> It's overall a bit odd that the smalltag is handled outside of `gc_alloc` e.g. callers have to check `dt->smalltag` and then use `jl_set_typetag_of`.
> 
> Maybe `gc_permobj` should take a `jl_value_ty*` and then be passed a `jl_symbol_type`, so that the caller doesn't have to pre-shift the tag?

I got a segfault when I tried to pass `jl_symbol_type` from `mk_symbol`. My guess is that `jl_symbol_type` is not initialized by the time `mk_symbol` is called (and why the tag was used there instead)?

udesou committed:
Passing alignment as parameter and cleaning up the code

udesou commented:
@Keno and @vchuravy I've tried to refactor the code a bit more taking into account your feedback. I've added the alignment as parameter to `jl_gc_permobj` and cleaned up the code in `mk_symbol` calling `jl_gc_permobj` instead. Let me know what you think, and if it still looks confusing.

d-netto reviewed (commented):
> This PR lifts up some code that uses specific information about the stock GC (GC_OLD_MARKED in the header bits) into a separate function adding it to the GC interface.

Looks like `mk_symbol`still references `GC_OLD_MARKED`.

Is the original goal of removing all stock-GC-specific code from this function still being met in the latest version of this PR?

udesou commented:
> > This PR lifts up some code that uses specific information about the stock GC (GC_OLD_MARKED in the header bits) into a separate function adding it to the GC interface.
> 
> Looks like `mk_symbol`still references `GC_OLD_MARKED`.
> 
> Is the original goal of removing all stock-GC-specific code from this function still being met in the latest version of this PR?

It is. MMTk sets the log/tag bit by calling post alloc, which happens for every object after allocating a `permobj`, or happens via the [notify](https://github.com/udesou/julia/blob/c4e559b0177724f69fb090508e519af1d83e9315/src/gc-interface.h#L213-L220) functions. It's fine to still set the tag bits, MMTk does not check them.

d-netto commented:
Latest commit seems to be failing MMTk CI?

I did a rebase a few minutes ago, but would not expect it to make tests suddenly fail.

d-netto reviewed (approved):
LGTM

d-netto commented on src/gc-mmtk.c:
```diff
@@ -1028,7 +1028,7 @@ void *jl_gc_perm_alloc(size_t sz, int zero, unsigned align, unsigned offset)
+jl_value_t *jl_gc_permobj(size_t sz, void *ty, unsigned align) JL_NOTSAFEPOINT
 {
     const size_t allocsz = sz + sizeof(jl_taggedvalue_t);
     unsigned align = (sz == 0 ? sizeof(void*) : (allocsz <= sizeof(void*) * 2 ?
```
Actually, you are failing CI because you are trying to shadow the function parameter `align` here...

udesou replied:
Good that we have that MMTk CI ;) 
Sorry that I forgot about it - will push a fix.
